File: dpd.py
n, W = map(int, input().split())

items = []

for _ in range(n):
    items.append(tuple(map(int, input().split())))

# sol = [[-1 for _ in range(W + 1)] for __ in range(n + 1)]

sol = {}


def knapsack(W_, items_):
    if W_ == 0 or not items_:
        sol[(len(items_), W_)] = 0
        return 0

    weight, value = items_[0]

    if sol.get((len(items_), W_)) != -1 and sol.get((len(items_), W_)):
        return sol[(len(items_), W_)]


    print('first', '\n', items, "\n")

    print('second', '\n', items[1:], '\n')


    if weight > W_:
        sol[(len(items_), W_)] = knapsack(W_, items_[1:])
        return sol[(len(items_), W_)]
    else:
        sol[(len(items_), W_)] = max(value + knapsack(W_ - weight, items_[1:]), knapsack(W_, items_[1:]))
        return sol[(len(items_), W_)]


print(knapsack(W, items))

print(sol[(len(items), W)])

File: dpd.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

//typedef vector <pair<int, long long > > vecpair;

//long long sol[101][100001];

//long long sol[10100100];

//
//long long knapsack(int W, vecpair items) {
//
//    if (W == 0 || items.empty()) {
////        sol[items.size()][W] = 0;
////        sol[items.size() * W] = 0;
//        return 0;
//    }
//
//    int weight = items.front().first;
//    long long value = items.front().second;
//
//
////    if (sol[items.size() * W] != 0) {
//////        return sol[items.size()][W];
////return sol[items.size() * W];
////    }
//
//
//    vecpair new_vec;
//    new_vec = vecpair(items.begin() + 1, items.end());
//
//
//    if (weight > W) {
////        sol[items.size() * W] = knapsack(W, new_vec);
////        return sol[items.size() * W];
//        return knapsack(W, new_vec);
//
//    } else {
////        long long a1 = value + knapsack(W - weight, new_vec);
////        long long a2 = knapsack(W, new_vec);
//
//
////        sol[items.size() * W] = max(value + knapsack(W - weight, new_vec), knapsack(W, new_vec));
////        return sol[items.size() * W];
//        return max(value + knapsack(W - weight, new_vec), knapsack(W, new_vec));
//
//
//    }
//}

int main() {

    cin.sync_with_stdio(0); cin.tie(0);
//    vecpair items;
    int N, W;


    long long sol[10000];

    cin >> N >> W;

    for (int i = 0; i < N; i++) {

        long long w;
        int v;
        cin >> w >> v;

        for (int wt = W; wt >= w; wt--) {
            sol[wt] = max(sol[wt], sol[wt - w] + v);
        }



//        items.push_back(make_pair(w_, v));
    }
    cout << sol[W] << "\n";


//    cout << knapsack(W, items) << "\n";

    return 0;
}
